### ISSUE TITLE
MQTT Plugin not finishing and lots of log entries

### ISSUE THREAD

**Issue by AdmiralRaccoon** [jokob-sk/NetAlertX#639]
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues and I checked the docs https://github.com/jokob-sk/NetAlertX/tree/main/docs

### Current Behavior

The MQTT Plugin runs almost 30 Minutes and then quits with a timeout, although a timeout of 10 Minutes is set.
Also the topic changes are not visible in Home Assistant.

### Expected Behavior

With 24.3.19, everything was ok and sending MQTT Topics took only a couple of seconds. I thought is was because of a faulty upgrade, but I restartet from scratch and have the same issues

### Steps To Reproduce

It´s starting after the scheduled device scan, when the mqtt plugin starts

### app.conf

```python
# MQTT
#---------------------------
MQTT_RUN='on_notification'
MQTT_RUN__metadata="{\"function\": \"RUN\", \"events\": [\"test\"], \"type\": \"text.select\", \"default_value\": \"disabled\", \"options\": [\"disabled\", \"on_notification\", \"once\", \"schedule\", \"always_after_scan\", \"on_new_device\"], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"When to run\"}, {\"language_code\": \"es_es\", \"string\": \"Cuando ejecuta\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Enable sending notifications via <a target=\\\"_blank\\\" href=\\\"https://www.home-assistant.io/integrations/mqtt/\\\">MQTT</a> to your Home Assistance instance. Usually, <code>on_notification</code> is recommended. See the <a target=\\\"_blank\\\" href=\\\"https://github.com/jokob-sk/Pi.Alert/blob/main/docs/HOME_ASSISTANT.md\\\">PiAlert Home Assistant guide</a> for details.\"}, {\"language_code\": \"es_es\", \"string\": \"Habilitar el env\\u00edo de notificaciones a trav\\u00e9s de <a target=\\\"_blank\\\" href=\\\"https://www.home-assistant.io/integrations/mqtt/\\\">MQTT</a> a su Home Assistance.\"}]}"
MQTT_CMD='python3 /home/pi/pialert/front/plugins/_publisher_mqtt/mqtt.py devices={devices}'
MQTT_CMD__metadata="{\"function\": \"CMD\", \"type\": \"readonly\", \"default_value\": \"python3 /home/pi/pialert/front/plugins/_publisher_mqtt/mqtt.py devices={devices}\", \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"Command\"}, {\"language_code\": \"es_es\", \"string\": \"Comando\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Command to run\"}, {\"language_code\": \"es_es\", \"string\": \"Comando a ejecutar\"}]}"
MQTT_RUN_SCHD='*/5 * * * *'
MQTT_RUN_SCHD__metadata="{\"function\": \"RUN_SCHD\", \"type\": \"text\", \"default_value\": \"0 2 * * 3\", \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"Schedule\"}, {\"language_code\": \"es_es\", \"string\": \"Schedule\"}, {\"language_code\": \"de_de\", \"string\": \"Schedule\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Only enabled if you select <code>schedule</code> in the <a href=\\\"#MQTT_RUN\\\"><code>MQTT_RUN</code> setting</a>. Make sure you enter the schedule in the correct cron-like format (e.g. validate at <a href=\\\"https://crontab.guru/\\\" target=\\\"_blank\\\">crontab.guru</a>). For example entering <code>0 4 * * *</code> will run the scan after 4 am in the <a onclick=\\\"toggleAllSettings()\\\" href=\\\"#TIMEZONE\\\"><code>TIMEZONE</code> you set above</a>. Will be run NEXT time the time passes.\"}, {\"language_code\": \"es_es\", \"string\": \"Solo est\\u00e1 habilitado si selecciona <code>schedule</code> en la configuraci\\u00f3n <a href=\\\"#MQTT_RUN\\\"><code>MQTT_RUN</code></a>. Aseg\\u00farese de ingresar la programaci\\u00f3n en el formato similar a cron correcto (por ejemplo, valide en <a href=\\\"https://crontab.guru/\\\" target=\\\"_blank\\\">crontab.guru</a>). Por ejemplo, ingresar <code>0 4 * * *</code> ejecutar\\u00e1 el escaneo despu\\u00e9s de las 4 a.m. en el <a onclick=\\\"toggleAllSettings()\\\" href=\\\"#TIMEZONE\\\"><code>TIMEZONE</ c\\u00f3digo> que configur\\u00f3 arriba</a>. Se ejecutar\\u00e1 la PR\\u00d3XIMA vez que pase el tiempo.\"}, {\"language_code\": \"de_de\", \"string\": \"Nur aktiviert, wenn Sie <code>schedule</code> in der <a href=\\\"#MQTT_RUN\\\"><code>MQTT_RUN</code>-Einstellung</a> ausw\\u00e4hlen. Stellen Sie sicher, dass Sie den Zeitplan im richtigen Cron-\\u00e4hnlichen Format eingeben (z. B. validieren unter <a href=\\\"https://crontab.guru/\\\" target=\\\"_blank\\\">crontab.guru</a>). Wenn Sie beispielsweise <code>0 4 * * *</code> eingeben, wird der Scan nach 4 Uhr morgens in der <a onclick=\\\"toggleAllSettings()\\\" href=\\\"#TIMEZONE\\\"><code>TIMEZONE</ ausgef\\u00fchrt. Code> den Sie oben festgelegt haben</a>. Wird das N\\u00c4CHSTE Mal ausgef\\u00fchrt, wenn die Zeit vergeht.\"}]}"
MQTT_RUN_TIMEOUT=10
MQTT_RUN_TIMEOUT__metadata="{\"function\": \"RUN_TIMEOUT\", \"type\": \"integer\", \"default_value\": 10, \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"Run timeout\"}, {\"language_code\": \"es_es\", \"string\": \"Tiempo de espera de ejecuci\\u00f3n\"}, {\"language_code\": \"de_de\", \"string\": \"Wartezeit\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Maximum time in seconds to wait for the script to finish. If this time is exceeded the script is aborted.\"}, {\"language_code\": \"es_es\", \"string\": \"Tiempo m\\u00e1ximo en segundos para esperar a que finalice el script. Si se supera este tiempo, el script se cancela.\"}]}"
MQTT_BROKER='192.168.1.99'
MQTT_BROKER__metadata="{\"function\": \"BROKER\", \"type\": \"text\", \"default_value\": \"\", \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"MQTT broker URL\"}, {\"language_code\": \"es_es\", \"string\": \"URL del broker MQTT\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"MQTT host URL (do not include <code>http://</code> or <code>https://</code>).\"}, {\"language_code\": \"es_es\", \"string\": \"URL del host MQTT (no incluya <code>http://</code> o <code>https://</code>).\"}]}"
MQTT_PORT=1883
MQTT_PORT__metadata="{\"function\": \"PORT\", \"type\": \"integer\", \"default_value\": 1883, \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"MQTT broker port\"}, {\"language_code\": \"es_es\", \"string\": \"Puerto del broker MQTT\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Port number where the broker is listening. Usually <code>1883</code>.\"}, {\"language_code\": \"es_es\", \"string\": \"Puerto donde escucha el broker MQTT. Normalmente <code>1883</code>.\"}]}"
MQTT_USER='mqttuser'
MQTT_USER__metadata="{\"function\": \"USER\", \"type\": \"text\", \"default_value\": \"\", \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"MQTT user\"}, {\"language_code\": \"es_es\", \"string\": \"Usuario de MQTT\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"User name used to login into your MQTT broker instance.\"}, {\"language_code\": \"es_es\", \"string\": \"Nombre de usuario utilizado para iniciar sesi\\u00f3n en su instancia de agente de MQTT.\"}]}"
MQTT_PASSWORD='mqttuser'
MQTT_PASSWORD__metadata="{\"function\": \"PASSWORD\", \"type\": \"password\", \"default_value\": \"\", \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"MQTT password\"}, {\"language_code\": \"es_es\", \"string\": \"Contrase\\u00f1a de MQTT\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Password used to login into your MQTT broker instance.\"}, {\"language_code\": \"es_es\", \"string\": \"Contrase\\u00f1a utilizada para iniciar sesi\\u00f3n en su instancia de agente de MQTT.\"}]}"
MQTT_QOS=0
MQTT_QOS__metadata="{\"function\": \"QOS\", \"type\": \"integer.select\", \"default_value\": 0, \"options\": [0, 1, 2], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"MQTT Quality of Service\"}, {\"language_code\": \"es_es\", \"string\": \"Calidad de servicio MQTT\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Quality of service setting for MQTT message sending. The higher the quality the longer the delay. <br/> <code>0</code> - Low quality to <code>2</code> - High quality.\"}, {\"language_code\": \"es_es\", \"string\": \"Configuraci\\u00f3n de calidad de servicio para el env\\u00edo de mensajes MQTT. <code>0</code>: baja calidad a <code>2</code>: alta calidad. Cuanto mayor sea la calidad, mayor ser\\u00e1 el retraso.\"}]}"
MQTT_VERSION=1
MQTT_VERSION__metadata="{\"function\": \"VERSION\", \"type\": \"integer.select\", \"default_value\": 1, \"options\": [1, 2], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"Version\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Paho MQTT API version. Depends on the MQTT <a href=\\\"https://eclipse.dev/paho/files/paho.mqtt.python/html/index.html#callbacks\\\" target=\\\"_blank\\\">version supported by the MQTT broker</a>. Usually set to <code>1</code>.\"}]}"
MQTT_DELAY_SEC=2
MQTT_DELAY_SEC__metadata="{\"function\": \"DELAY_SEC\", \"type\": \"integer\", \"default_value\": 2, \"options\": [2, 3, 4, 5], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"MQTT delay per device\"}, {\"language_code\": \"es_es\", \"string\": \"Retraso de MQTT por dispositivo\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"A little hack - delay adding to the queue in case the process is restarted and previous publish processes aborted (it takes ~<code>2</code>s to update a sensor config on the broker). Tested with <code>2</code>-<code>3</code> seconds of delay. This delay is only applied when devices are created (during the first notification loop). It doesn not affect subsequent scans or notifications.\"}, {\"language_code\": \"es_es\", \"string\": \"Un peque\\u00f1o truco: retrase la adici\\u00f3n a la cola en caso de que el proceso se reinicie y los procesos de publicaci\\u00f3n anteriores se anulen (se necesitan ~<code>2</code>s para actualizar la configuraci\\u00f3n de un sensor en el intermediario). Probado con <code>2</code>-<code>3</code> segundos de retraso. Este retraso solo se aplica cuando se crean dispositivos (durante el primer bucle de notificaci\\u00f3n). No afecta los escaneos o notificaciones posteriores.\"}]}"
MQTT_SEND_STATS=True
MQTT_SEND_STATS__metadata="{\"function\": \"SEND_STATS\", \"type\": \"boolean\", \"default_value\": true, \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"Send stats\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Check to send overal device stats, such as number of Online and Offline devices.\"}]}"
MQTT_SEND_DEVICES=True
MQTT_SEND_DEVICES__metadata="{\"function\": \"SEND_DEVICES\", \"type\": \"boolean\", \"default_value\": true, \"options\": [], \"localized\": [\"name\", \"description\"], \"name\": [{\"language_code\": \"en_us\", \"string\": \"Send devices\"}], \"description\": [{\"language_code\": \"en_us\", \"string\": \"Check to send individual devices to the broker with details, such as <code>is_new</code>, <code>is_present</code>, or <code>mac_address</code> of the devices.\"}]}"
```

### docker-compose.yml

```python
version: "3"
services:
  netalertx:                                  # ⚠  This has changed (🟡optional) 
    container_name: netalertx                 # ⚠  This has changed (🟡optional) 
    # use the below line if you want to test the latest dev image
    # image: "jokobsk/netalertx-dev:latest" 
    image: "jokobsk/netalertx:latest"         # ⚠  This has changed (🟡optional/🔺required in future) 
    network_mode: "host"        
    restart: always
    volumes:
      - /netalertx/config:/app/config         # ⚠  This has changed (🔺required) 
      - /netalertx/db:/app/db                 # ⚠  This has changed (🔺required) 
      # (optional) useful for debugging if you have issues setting up the container
      - /netalertx/logs:/app/front/log        # ⚠  This has changed (🟡optional) 
    environment:
      - TZ=Europe/Berlin      
      - PORT=20211
```

### What branch are you running?

Production

### app.log

this is a small part of the log after i restartet the container. if you need more, please tell me.

[app.log](https://github.com/jokob-sk/NetAlertX/files/15048659/app.log)

### Debug enabled

- [X] I have read and followed the steps in the wiki link above and provided the required debug logs and the log section covers the time when the issue occurs.

**Comment by jokob-sk** [comment]
Hi @AdmiralRaccoon ,

Thanks for the report.

I couldn't see any error in the event log so if you have the specific timeout or other error you could provide, that would help a lot.

```
16:46:21 [MQTT]         Estimated delay: 1460s (24.3min)
```

According to your log, the sync is estimated to take around 24 min. This time is an estimation based on the timeout setting (in seconds) multiplied by the number of devices. If you want to abort the script earlier, decrease this setting (default is 10s per device).

![image](https://github.com/jokob-sk/NetAlertX/assets/96159884/7237bff0-59f5-486c-8fc6-d3ecfc58cdd9)

Another setting you can tweak is `MQTT delay per device`. You can decrease it to 1s (2s by default). 

In general however, the first run will take a while, subsequent should be much quicker. I'll also explain why:

1. The app keeps a hash of the sensors and their state
2. This hash is compared before a MQTT message is sent to the broker
3. If the hash is not found, a new device/device state is assumed and the device is sent to the broker

The hash can be found under Integrations > Plugins > MQTT (Plugin objects tab). 

The rename from PiALert to NetAlertX will have triggered a refresh of the sensors, so I assume that's why the initial MQTT sync on the new build takes longer than the previous MQTT updates on an already running instance. 

I also will be adding 2 more settings to change the DeviceId/Name root in the messages to give more configuration freedom in the next release:

![image](https://github.com/jokob-sk/NetAlertX/assets/96159884/99edaac3-f0bf-4275-907c-7c9a93893777)

This should be available in the next release. If you can, please have a look at the `netalertx-dev` docker image, in about 15 minutes (or after the last [action finishes](https://github.com/jokob-sk/NetAlertX/actions)) from now. 

In summary, try extending / decreasing the timeouts to get the first MQTT run finished and hopefully subsequent ones will be much faster. 

hope this helps,
j

**Comment by AdmiralRaccoon** [comment]
Good Morning and thanks for the detailed reply, I will look into it during the day.

Although i enabled the debug log level, there is not much information when the mqtt timeout occurs:

```
09:45:55 [Plugins] ⚠ ERROR - TIMEOUT - the plugin MQTT forcefully terminated as timeout reached. Increase TIMEOUT setting and scan interval.
09:45:55 [Plugins] The file /app/front/plugins/_publisher_mqtt/last_result.log does not exist
09:45:55 [Plugins] No output received from the plugin MQTT - enable LOG_LEVEL=debug and check logs`
```

**Comment by jokob-sk** [comment]
Thanks for the details. In that case, I would try increasing the `MQTT_RUN_TIMEOUT` to e.g. 30s per device. I hope this will give it enough time to push the changes thru.🤞 Let me know how you go.

**Comment by AdmiralRaccoon** [comment]
1. I´m trying to adjust the settings, but it´s difficult to do so. when the plugin: mqtt is running, is the config somehow locked, because I only get this Message the whole time?

![image](https://github.com/jokob-sk/NetAlertX/assets/41727850/e5190c36-20d5-403a-83b9-2ac7d4bfba7e)


EDIT: I restartet the container, quickly disabled mqtt, set the timeout to 30 but the result is pretty the same. the Plugin is so slow, it takes hours for home assistant to see that a device goes online/offline

Just curious - With the March Release I had no MQTT issues at all, I used it as a device tracker in Home Assistant and it was pretty fast. Was there a significant change in the April release maybe?




2. I´ve also tried to test the dev image, but for now i can´t access the settings because it stops at reading the config

```
07:08:30 The container restarted (started). If this is unexpected check https://bit.ly/NetAlertX_debug for troubleshooting tips.
07:08:30 

07:08:30 Permissions check (All should be True)
07:08:30 ------------------------------------------------
07:08:30   /config/app.conf |  READ  | True
07:08:30   /config/app.conf |  WRITE | True
07:08:30   /db/app.db       |  READ  | True
07:08:30   /db/app.db       |  WRITE | True
07:08:30 ------------------------------------------------
07:08:30 [Setup] Attempting to fix permissions.
07:08:30 [Setup] Attempting to fix permissions.
07:08:30 [Database] Opening DB
07:08:30 [upgradeDB] Re-creating Settings table
07:08:30 [upgradeDB] Re-creating Pholus_Scan table
07:08:30 [upgradeDB] Re-creating Parameters table
07:08:30 [Config] reading config file
```

**Comment by AdmiralRaccoon** [comment]
Ok hold on, right now it seems to work, the mqtt plugin seems to finish gracefully - although i havn´t done much yet (except the 30 seconds timeout)
I´ll keep you informed, thanks for everything!

**Comment by jokob-sk** [comment]
Glad to hear there is progress! I increased the default to 20s so hopefully, that will prevent similar issues in the future.
